In [1]:
import pandas as pd
import re

In [2]:
df = pd.read_excel('../data/Customer Call List.xlsx')
df.head(3)

CustomerID First_Name Last_Name  Phone_Number                Address  \
0        1001      Frodo   Baggins  123-545-5421  123 Shire Lane, Shire   
1        1002       Abed     Nadir  123/643/9775    93 West Main Street   
2        1003     Walter    /White    7066950392     298 Drugs Driveway   

  Paying Customer Do_Not_Contact  Not_Useful_Column  
0             Yes             No               True  
1              No            Yes              False  
2               N            NaN               True

In [3]:
# df.dtypes
df.shape

(21, 8)

### Data Format Issues:

- Phone_Number: The format is inconsistent with a mix of hyphens, slashes, and pipes as delimiters.
- Address: Some addresses lack zip codes or have state information embedded within the address field.
- Paying Customer: Values are inconsistent, using both "Yes/No" and "Y/N" variations.
- Do_Not_Contact: Similar to "Paying Customer," uses both "Yes/No" and "Y/N" variations. Also, some entries are blank, which may need to be treated as a separate category or standardized.
- Not_Useful_Column: This column seems irrelevant and can be removed.

### Data Quality Issues:
- Duplicate Entry: CustomerID 1020 appears twice with identical information.
- Missing Values: Several fields have missing values, notably in Last_Name for customer 1003 and Phone_Number for customer 1007.
- Inconsistent Naming: First and last names sometimes include extra spaces (e.g., Jeff Winger, Michael Scott).

### Recommendations:
#### 1. Data Cleaning:
- Standardize phone number format.
- Separate state and zip code in the address field.
- Unify "Paying Customer" and "Do_Not_Contact" values to either "Yes/No" or "Y/N".
- Address missing values through further investigation or appropriate imputation methods.
- Remove extra spaces from names and ensure consistent capitalization.
- Remove duplicate entry for customer 1020.
- Delete the "Not_Useful_Column".

#### 2. Data Validation:
- Implement data validation rules to prevent future inconsistencies.
- For instance, enforce format restrictions on phone numbers and ensure mandatory fields are filled.

#### Additional Notes:
- Consider the purpose of the data and how it will be used. This will guide further cleaning and analysis steps.
- Explore potential relationships between variables (e.g., Paying Customer and Do_Not_Contact).
- Depending on the intended use, you might explore data enrichment by appending additional information such as email addresses or demographic data.

By addressing these issues, you can improve the quality and reliability of the customer data for better decision-making and analysis.

In [4]:
# # Steps
# remove duplicates
# drop columns
# standardize the data
# Null values or blank values
# remove any columns

In [5]:
# create a copy of the dataframe
df_copy = df.copy()
df_copy.head(3)

CustomerID First_Name Last_Name  Phone_Number                Address  \
0        1001      Frodo   Baggins  123-545-5421  123 Shire Lane, Shire   
1        1002       Abed     Nadir  123/643/9775    93 West Main Street   
2        1003     Walter    /White    7066950392     298 Drugs Driveway   

  Paying Customer Do_Not_Contact  Not_Useful_Column  
0             Yes             No               True  
1              No            Yes              False  
2               N            NaN               True

In [6]:
# check for duplicates and show the rows
df_copy[df_copy.duplicated()]

CustomerID First_Name  Last_Name  Phone_Number  \
20        1020     Anakin  Skywalker  876|678|3469   

                        Address Paying Customer Do_Not_Contact  \
20  910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
20               True

In [7]:
# remove duplicates
df_copy.drop_duplicates(inplace=True)

# preview df
df_copy.head(3)

CustomerID First_Name Last_Name  Phone_Number                Address  \
0        1001      Frodo   Baggins  123-545-5421  123 Shire Lane, Shire   
1        1002       Abed     Nadir  123/643/9775    93 West Main Street   
2        1003     Walter    /White    7066950392     298 Drugs Driveway   

  Paying Customer Do_Not_Contact  Not_Useful_Column  
0             Yes             No               True  
1              No            Yes              False  
2               N            NaN               True

In [8]:
# drop columns
df_copy.drop(columns= 'Not_Useful_Column')

# preview df
df_copy

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True

In [9]:
# standardize the data "Last_Name" by removing white space
# df_copy["Last_Name"].str.strip() # remove white space by default
# df_copy["Last_Name"] = df_copy["Last_Name"].str.lstrip("...")
# df_copy["Last_Name"] = df_copy["Last_Name"].str.lstrip("/") 
# df_copy["Last_Name"] = df_copy["Last_Name"].str.rstrip("_")

df_copy['Last_Name'] = df_copy['Last_Name'].str.strip("123._/")
df_copy

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123/643/9775   
2         1003     Walter       White    7066950392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876|678|3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger           NaN   
7         1008   Sherlock      Holmes  876|678|3469   
8         1009    Gandalf         NaN           N/a   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee           NaN   
11        1012      Harry      Potter    7066950392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876|678|3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123/643/9775   
17        1018      Clark        Kent    7066950392   
18        1019      Creed      Braton           N/a   
19        1020     Anakin   Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True

In [10]:
df_copy['Phone_Number']

0     123-545-5421
1     123/643/9775
2       7066950392
3     123-543-2345
4     876|678|3469
5     304-762-2467
6              NaN
7     876|678|3469
8              N/a
9     123-545-5421
10             NaN
11      7066950392
12    123-543-2345
13    876|678|3469
14    304-762-2467
15    123-545-5421
16    123/643/9775
17      7066950392
18             N/a
19    876|678|3469
Name: Phone_Number, dtype: object

In [11]:
# Remove non-numeric characters from the phone numbers
df_copy['Phone_Number'] = df_copy['Phone_Number'].str.replace(r'[^0-9]+', '', regex=True)

# Add hyphens to format the phone numbers as XXX-XXX-XXXX
df_copy['Phone_Number'] = df_copy['Phone_Number'].str.replace(r'(\d{3})(\d{3})(\d{4})', r'\1-\2-\3')

# Display the standardized phone numbers
print(df_copy['Phone_Number'])

0     1235455421
1     1236439775
2            NaN
3     1235432345
4     8766783469
5     3047622467
6            NaN
7     8766783469
8               
9     1235455421
10           NaN
11           NaN
12    1235432345
13    8766783469
14    3047622467
15    1235455421
16    1236439775
17           NaN
18              
19    8766783469
Name: Phone_Number, dtype: object
